File: jav/111_minimumDepthOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;

        int minDepth = Integer.MAX_VALUE;
        int curDepth = 0;
        return minDepthHelper(root, minDepth, curDepth);
    }

    public int minDepthHelper(TreeNode root, int minDepth, int curDepth)
    {
        if (root == null)
        {
            return Integer.MAX_VALUE;
        }
        else
        {
            if (root.left == null && root.right == null)
            {
                return Math.min(minDepth, curDepth + 1);
            }
            else
            {
                int leftMin = minDepthHelper(root.left, minDepth, curDepth + 1);
                int rightMin = minDepthHelper(root.right, minDepth, curDepth + 1);
                return Math.min(leftMin, rightMin);
            }
        }
    }
}

File: cpp/111_minimumDepthOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        int minDepth = INT_MAX;
        int curDepth = 0;
        if (!root) return 0;
        minDepthHelper(root, minDepth, curDepth);
        return minDepth;
    }
    
    void minDepthHelper(TreeNode* root, int& minDepth, int& curDepth)
    {
        if (root)
        {
            curDepth += 1;
            
            if (!root->left && !root->right)
            {
                minDepth = std::min(minDepth, curDepth);
            }
            else
            {
                if (root->left) minDepthHelper(root->left, minDepth, curDepth);
                if (root->right) minDepthHelper(root->right, minDepth, curDepth);
            }
            
        curDepth -= 1;
        }
    }
};